File: lexer.js
class Lexer {
    constructor() {
        this.keywords = new Set([
            "auto", "break", "case", "char", "const", "continue", "default", "do",
            "double", "else", "enum", "extern", "float", "for", "goto", "if",
            "int", "long", "register", "return", "short", "signed", "sizeof", "static",
            "struct", "switch", "typedef", "union", "unsigned", "void", "volatile", "while"
        ]);

        this.operators = new Set([
            "+", "-", "*", "/", "%", "=", "==", "!=", "<", ">", "<=", ">=",
            "&&", "||", "!", "&", "|", "^", "~", "<<", ">>", "+=", "-=", "*=",
            "/=", "%=", "&=", "|=", "^=", "<<=", ">>=", "++", "--", "->", "."
        ]);

        this.punctuation = new Set([
            ";", ",", "(", ")", "{", "}", "[", "]", "#", ":"
        ]);

        this.tokens = [];
        this.statistics = {
            keywords: 0,
            identifiers: 0,
            operators: 0,
            constants: 0,
            punctuation: 0,
            strings: 0,
            preprocessor: 0
        };
    }

    // Remove comments (both single-line and multi-line)
    removeComments(source) {
        let result = "";
        let inSingleLineComment = false;
        let inMultiLineComment = false;
        let inStringLiteral = false;
        let inCharLiteral = false;

        for (let i = 0; i < source.length; i++) {
            const char = source[i];
            const nextChar = i < source.length - 1 ? source[i + 1] : '';

            // Handle string literals (don't remove comments inside strings)
            if (char === '"' && !inCharLiteral && !inSingleLineComment && !inMultiLineComment) {
                if (i === 0 || source[i - 1] !== '\\') {
                    inStringLiteral = !inStringLiteral;
                }
                result += char;
                continue;
            }

            // Handle character literals
            if (char === "'" && !inStringLiteral && !inSingleLineComment && !inMultiLineComment) {
                if (i === 0 || source[i - 1] !== '\\') {
                    inCharLiteral = !inCharLiteral;
                }
                result += char;
                continue;
            }

            // Skip comment processing if inside string or char literal
            if (inStringLiteral || inCharLiteral) {
                result += char;
                continue;
            }

            if (inSingleLineComment) {
                if (char === '\n') {
                    inSingleLineComment = false;
                    result += '\n';
                }
                continue;
            }

            if (inMultiLineComment) {
                if (char === '*' && nextChar === '/') {
                    inMultiLineComment = false;
                    i++; // Skip the '/'
                }
                continue;
            }

            if (char === '/' && nextChar === '/') {
                inSingleLineComment = true;
                i++; // Skip the second '/'
                continue;
            }

            if (char === '/' && nextChar === '*') {
                inMultiLineComment = true;
                i++; // Skip the '*'
                continue;
            }

            result += char;
        }

        return result;
    }

    // Remove extra whitespace while preserving structure
    removeWhitespace(source) {
        let result = "";
        let lastWasSpace = true;

        for (const char of source) {
            if (/\s/.test(char)) {
                if (char === '\n') {
                    result += '\n';
                    lastWasSpace = true;
                } else if (!lastWasSpace) {
                    result += ' ';
                    lastWasSpace = true;
                }
            } else {
                result += char;
                lastWasSpace = false;
            }
        }

        return result;
    }

    // Tokenize the source code
    tokenize(source) {
        this.tokens = [];
        this.resetStatistics();
        
        let token = "";
        let lineNumber = 1;
        let inStringLiteral = false;
        let inCharLiteral = false;

        for (let i = 0; i < source.length; i++) {
            const char = source[i];
            const nextChar = i < source.length - 1 ? source[i + 1] : '';

            // Track line numbers
            if (char === '\n') {
                lineNumber++;
            }

            // Handle string literals
            if (char === '"' && !inCharLiteral) {
                if (token !== "" && !inStringLiteral) {
                    this.categorizeToken(token, lineNumber);
                    token = "";
                }
                
                if (!inStringLiteral) {
                    // Start of string
                    let stringToken = '"';
                    i++;
                    while (i < source.length && (source[i] !== '"' || source[i-1] === '\\')) {
                        stringToken += source[i];
                        if (source[i] === '\n') lineNumber++;
                        i++;
                    }
                    if (i < source.length) stringToken += '"';
                    
                    this.addToken('String Literal', stringToken, lineNumber);
                    this.statistics.strings++;
                } 
                continue;
            }

            // Handle character literals
            if (char === "'" && !inStringLiteral) {
                if (token !== "") {
                    this.categorizeToken(token, lineNumber);
                    token = "";
                }
                
                let charToken = "'";
                i++;
                while (i < source.length && (source[i] !== "'" || source[i-1] === '\\')) {
                    charToken += source[i];
                    i++;
                }
                if (i < source.length) charToken += "'";
                
                this.addToken('Character Literal', charToken, lineNumber);
                this.statistics.constants++;
                continue;
            }

            // Handle preprocessor directives
            if (char === '#' && (i === 0 || source[i-1] === '\n')) {
                if (token !== "") {
                    this.categorizeToken(token, lineNumber);
                    token = "";
                }
                
                let preprocessor = "#";
                i++;
                while (i < source.length && source[i] !== '\n') {
                    preprocessor += source[i];
                    i++;
                }
                
                this.addToken('Preprocessor', preprocessor.trim(), lineNumber);
                this.statistics.preprocessor++;
                continue;
            }

            // Handle punctuation and operators
            if (this.isPunctuation(char) || this.couldBeOperator(char)) {
                if (token !== "") {
                    this.categorizeToken(token, lineNumber);
                    token = "";
                }

                // Check for multi-character operators
                let operator = char;
                if (this.operators.has(char + nextChar)) {
                    operator = char + nextChar;
                    i++;
                } else if (i < source.length - 2 && this.operators.has(char + nextChar + source[i + 2])) {
                    operator = char + nextChar + source[i + 2];
                    i += 2;
                }

                if (this.operators.has(operator)) {
                    this.addToken('Operator', operator, lineNumber);
                    this.statistics.operators++;
                } else if (this.punctuation.has(operator)) {
                    this.addToken('Punctuation', operator, lineNumber);
                    this.statistics.punctuation++;
                }
                continue;
            }

            // Handle whitespace
            if (/\s/.test(char)) {
                if (token !== "") {
                    this.categorizeToken(token, lineNumber);
                    token = "";
                }
                continue;
            }

            token += char;
        }

        // Handle the last token
        if (token !== "") {
            this.categorizeToken(token, lineNumber);
        }

        return this.tokens;
    }

    // Helper methods
    isPunctuation(char) {
        return this.punctuation.has(char);
    }

    couldBeOperator(char) {
        return /[+\-*\/%=<>!&|^~]/.test(char);
    }

    // Categorize individual tokens
    categorizeToken(token, lineNumber) {
        // Check if it's a keyword
        if (this.keywords.has(token)) {
            this.addToken('Keyword', token, lineNumber);
            this.statistics.keywords++;
            return;
        }

        // Check if it's a numeric constant
        if (/^[0-9]+(\.[0-9]+)?([eE][+-]?[0-9]+)?$/.test(token) || 
            /^0[xX][0-9a-fA-F]+$/.test(token) || 
            /^0[0-7]+$/.test(token)) {
            this.addToken('Numeric Constant', token, lineNumber);
            this.statistics.constants++;
            return;
        }

        // Check if it's an identifier
        if (/^[a-zA-Z_][a-zA-Z0-9_]*$/.test(token)) {
            this.addToken('Identifier', token, lineNumber);
            this.statistics.identifiers++;
            return;
        }

        // If none of the above, it might be an unrecognized token
        this.addToken('Unknown', token, lineNumber);
    }

    addToken(type, value, lineNumber) {
        this.tokens.push({
            type: type,
            value: value,
            line: lineNumber
        });
    }

    resetStatistics() {
        this.statistics = {
            keywords: 0,
            identifiers: 0,
            operators: 0,
            constants: 0,
            punctuation: 0,
            strings: 0,
            preprocessor: 0
        };
    }

    // Main analysis function
    analyze(sourceCode) {
        // Step 1: Remove comments
        const withoutComments = this.removeComments(sourceCode);
        
        // Step 2: Tokenize
        const tokens = this.tokenize(sourceCode); // Use original source to preserve line numbers
        
        // Step 3: Clean for display
        const cleanedCode = this.removeWhitespace(withoutComments);

        return {
            tokens: tokens,
            processedCode: cleanedCode,
            statistics: this.statistics,
            originalWithoutComments: withoutComments
        };
    }
}

// Export for use in other files
if (typeof module !== 'undefined' && module.exports) {
    module.exports = Lexer;
}

File: src/lexer.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <regex>
#include <set>

class Lexer {
private:
    std::set<std::string> keywords = {
        "auto", "break", "case", "char", "const", "continue", "default", "do",
        "double", "else", "enum", "extern", "float", "for", "goto", "if",
        "int", "long", "register", "return", "short", "signed", "sizeof", "static",
        "struct", "switch", "typedef", "union", "unsigned", "void", "volatile", "while"
    };

    std::set<std::string> operators = {
        "+", "-", "*", "/", "%", "=", "==", "!=", "<", ">", "<=", ">=",
        "&&", "||", "!", "&", "|", "^", "~", "<<", ">>", "+=", "-=", "*=",
        "/=", "%=", "&=", "|=", "^=", "<<=", ">>=", "++", "--", "->", "."
    };

public:
    // Remove comments (both single-line and multi-line)
    std::string removeComments(const std::string& source) {
        std::string result;
        bool inSingleLineComment = false;
        bool inMultiLineComment = false;
        
        for (size_t i = 0; i < source.length(); ++i) {
            if (inSingleLineComment) {
                if (source[i] == '\n') {
                    inSingleLineComment = false;
                    result += '\n';
                }
                continue;
            }
            
            if (inMultiLineComment) {
                if (i < source.length() - 1 && source[i] == '*' && source[i + 1] == '/') {
                    inMultiLineComment = false;
                    ++i;
                }
                continue;
            }
            
            if (i < source.length() - 1) {
                if (source[i] == '/' && source[i + 1] == '/') {
                    inSingleLineComment = true;
                    ++i;
                    continue;
                }
                if (source[i] == '/' && source[i + 1] == '*') {
                    inMultiLineComment = true;
                    ++i;
                    continue;
                }
            }
            
            if (!inSingleLineComment && !inMultiLineComment) {
                result += source[i];
            }
        }
        return result;
    }

    // Remove extra whitespace while preserving newlines
    std::string removeWhitespace(const std::string& source) {
        std::string result;
        bool lastWasSpace = true;  // Start true to trim leading spaces
        
        for (char c : source) {
            if (std::isspace(c)) {
                if (c == '\n') {
                    result += '\n';
                    lastWasSpace = true;
                } else if (!lastWasSpace) {
                    result += ' ';
                    lastWasSpace = true;
                }
            } else {
                result += c;
                lastWasSpace = false;
            }
        }
        return result;
    }

    // Recognize and categorize tokens
    void tokenize(const std::string& source) {
        std::string token;
        std::string currentLine;
        
        for (size_t i = 0; i < source.length(); ++i) {
            char c = source[i];
            
            // Handle operators
            if (ispunct(c)) {
                if (!token.empty()) {
                    categorizeToken(token);
                    token.clear();
                }
                
                std::string op;
                op += c;
                if (i + 1 < source.length()) {
                    std::string doubleOp = op + source[i + 1];
                    if (operators.find(doubleOp) != operators.end()) {
                        op = doubleOp;
                        ++i;
                    }
                }
                
                if (operators.find(op) != operators.end()) {
                    std::cout << "Operator: " << op << std::endl;
                }
                continue;
            }

            // Handle whitespace
            if (std::isspace(c)) {
                if (!token.empty()) {
                    categorizeToken(token);
                    token.clear();
                }
                if (c == '\n') {
                    std::cout << std::endl;
                }
                continue;
            }

            token += c;
        }
        
        // Handle the last token
        if (!token.empty()) {
            categorizeToken(token);
        }
    }

private:
    // Categorize individual tokens
    void categorizeToken(const std::string& token) {
        // Check if it's a keyword
        if (keywords.find(token) != keywords.end()) {
            std::cout << "Keyword: " << token << std::endl;
            return;
        }

        // Check if it's a constant (number)
        if (std::regex_match(token, std::regex("^[0-9]+(\\.[0-9]+)?$"))) {
            std::cout << "Constant: " << token << std::endl;
            return;
        }

        // Check if it's an identifier
        if (std::regex_match(token, std::regex("^[a-zA-Z_][a-zA-Z0-9_]*$"))) {
            std::cout << "Identifier: " << token << std::endl;
            return;
        }
    }
};

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <source_file>" << std::endl;
        return 1;
    }

    std::ifstream file(argv[1]);
    if (!file.is_open()) {
        std::cerr << "Error: Could not open file " << argv[1] << std::endl;
        return 1;
    }

    // Read the entire file into a string
    std::string source((std::istreambuf_iterator<char>(file)),
                       std::istreambuf_iterator<char>());
    file.close();

    Lexer lexer;
    
    // Process the source code
    std::cout << "Processing file: " << argv[1] << "\n\n";
    
    // Remove comments first
    source = lexer.removeComments(source);
    
    // Remove extra whitespace
    source = lexer.removeWhitespace(source);
    
    // Tokenize and categorize
    std::cout << "Tokens:\n";
    std::cout << "-------\n";
    lexer.tokenize(source);

    return 0;
}